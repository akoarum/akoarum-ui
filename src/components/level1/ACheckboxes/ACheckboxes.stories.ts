import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withKnobs } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import ACheckboxes from './ACheckboxes.vue'

storiesOf('level1/ACheckboxes', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Normal', () => ({
    components: { ACheckboxes },
    data() {
      return {
        value: [],
        items: [
          { id: 0, label: '項目1', value: 1 },
          { id: 1, label: '項目2', value: 2 },
          { id: 2, label: '項目3', value: 3 }
        ]
      }
    },
    template: `
      <ACheckboxes v-model="value" :items="items" name="checkbox" />
    `
  }), { info: {} })
  .add('Error', () => ({
    components: { ACheckboxes },
    data() {
      return {
        value: [],
        items: [
          { id: 0, label: '項目1', value: 1 },
          { id: 1, label: '項目2', value: 2 },
          { id: 2, label: '項目3', value: 3 }
        ]
      }
    },
    template: `
      <ACheckboxes v-model="value" :items="items" name="checkbox" aria-invalid="true" />
    `
  }), { info: {} })
