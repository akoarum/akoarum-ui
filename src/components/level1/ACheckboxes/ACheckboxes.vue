<template>
  <ul class="aCheckboxes">
    <template v-for="item in items">
      <li :key="item.id" class="aCheckboxes__item">
        <label class="aCheckboxes__label">
          <input
            v-model="valueModel"
            v-bind="$attrs"
            :name="name"
            :value="item.value"
            type="checkbox"
            class="aCheckboxes__input"
          />
          <span class="aCheckboxes__text">{{ item.label }}</span>
        </label>
      </li>
    </template>
  </ul>
</template>

<script lang="ts">
import { defineComponent, computed, SetupContext } from '@vue/composition-api'
import { FormOption } from '~/types/FormOption'

interface Props {
  name: string
  items: FormOption[]
  value: (string | number)[]
}

export default defineComponent({
  inheritAttrs: false,
  model: {
    event: 'change'
  },
  props: {
    name: { type: String, required: true },
    items: { type: Array, required: true },
    value: { type: Array, required: true }
  },
  setup(props: Props, context: SetupContext) {
    const valueModel = computed({
      get: (): (string | number)[] => props.value,
      set: (newValue: (string | number)[]) => {
        context.emit('change', newValue)
      }
    })

    return { valueModel }
  }
})
</script>

<style lang="scss" scoped>
.aCheckboxes__item {
  font-size: 15px;
}

.aCheckboxes__item + .aCheckboxes__item {
  margin-top: 8px;
}

.aCheckboxes__label {
  position: relative;
  display: inline-block;
  padding-left: 24px;
}

.aCheckboxes__input {
  position: absolute;
  left: 0;
  top: 0;
  width: 4px;
  height: 4px;
  opacity: 0;

  &:checked + .aCheckboxes__text {
    &::before {
      border-color: $color-primary;
      background: $color-primary;
    }

    &::after {
      opacity: 1;
    }
  }
  &[aria-invalid] + .aCheckboxes__text {
    color: $color-caution;

    &::before {
      border-color: $color-caution;
      background: #fff7f7;
    }

    &::after {
      background: $color-caution;
    }
  }
}

.aCheckboxes__text {
  &::before,
  &::after {
    position: absolute;
    content: '';
  }

  &::before {
    top: 2px;
    left: 0;
    border: 2px solid $color-base;
    width: 14px;
    height: 14px;
    background: #fff;
    transition: border $duration $easing, background $duration $easing;
  }

  &::after {
    top: 3px;
    left: 1px;
    content: 'check';
    font-family: 'Material Icons';
    font-size: 16px;
    color: #fff;
    opacity: 0;
    transition: opacity $duration $easing;
  }
}
</style>
