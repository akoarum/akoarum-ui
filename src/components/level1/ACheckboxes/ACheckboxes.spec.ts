import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import ACheckboxes from './ACheckboxes.vue'

describe('ACheckboxes', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(ACheckboxes, {
      propsData: {
        name: 'test',
        value: [],
        items: [
          { id: 0, label: '項目1', value: 1 },
          { id: 1, label: '項目2', value: 2 },
          { id: 2, label: '項目3', value: 3 }
        ]
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
