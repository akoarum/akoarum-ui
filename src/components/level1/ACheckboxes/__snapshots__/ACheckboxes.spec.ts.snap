// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ACheckboxes initialized snapshot 1`] = `
<ul
  class="aCheckboxes"
>
  <li
    class="aCheckboxes__item"
  >
    <label
      class="aCheckboxes__label"
    >
      <input
        class="aCheckboxes__input"
        name="test"
        type="checkbox"
        value="1"
      />
       
      <span
        class="aCheckboxes__text"
      >
        項目1
      </span>
    </label>
  </li>
  <li
    class="aCheckboxes__item"
  >
    <label
      class="aCheckboxes__label"
    >
      <input
        class="aCheckboxes__input"
        name="test"
        type="checkbox"
        value="2"
      />
       
      <span
        class="aCheckboxes__text"
      >
        項目2
      </span>
    </label>
  </li>
  <li
    class="aCheckboxes__item"
  >
    <label
      class="aCheckboxes__label"
    >
      <input
        class="aCheckboxes__input"
        name="test"
        type="checkbox"
        value="3"
      />
       
      <span
        class="aCheckboxes__text"
      >
        項目3
      </span>
    </label>
  </li>
</ul>
`;
